core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+	{ FILTER_CONSTRAINT_AIR_TEMP, "air_temp", QT_TRANSLATE_NOOP("gettextFromC", "air temp."), false, true, false, FILTER_CONSTRAINT_TEMPERATURE_UNIT, 1, false, false },
+	{ FILTER_CONSTRAINT_SAC, "sac", QT_TRANSLATE_NOOP("gettextFromC", "SAC"), false, true, false, FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT, 1, false, false },
+
+	{ FILTER_CONSTRAINT_LOGGED, "logged", QT_TRANSLATE_NOOP("gettextFromC", "logged"), false, false, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+	{ FILTER_CONSTRAINT_PLANNED, "planned", QT_TRANSLATE_NOOP("gettextFromC", "planned"), false, false, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+
+	{ FILTER_CONSTRAINT_DIVE_MODE, "dive_mode", QT_TRANSLATE_NOOP("gettextFromC", "dive mode"), false, false, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+
+	{ FILTER_CONSTRAINT_TAGS, "tags", QT_TRANSLATE_NOOP("gettextFromC", "tags"), true, false, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+	{ FILTER_CONSTRAINT_PEOPLE, "people", QT_TRANSLATE_NOOP("gettextFromC", "people"), true, false, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+	{ FILTER_CONSTRAINT_LOCATION, "location", QT_TRANSLATE_NOOP("gettextFromC", "location"), true, false, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+	{ FILTER_CONSTRAINT_SUIT, "suit", QT_TRANSLATE_NOOP("gettextFromC", "suit"), true, false, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+	{ FILTER_CONSTRAINT_NOTES, "notes", QT_TRANSLATE_NOOP("gettextFromC", "notes"), true, false, false, FILTER_CONSTRAINT_NO_UNIT, 0, false, false },
+};
+
+static struct string_mode_description {
+	filter_constraint_string_mode mode;
+	const char *s;
+	const char *translated;
+} string_mode_descriptions[] = {
+	{ FILTER_CONSTRAINT_STARTS_WITH, "starts_with", QT_TRANSLATE_NOOP("gettextFromC", "starting with") },
+	{ FILTER_CONSTRAINT_SUBSTRING, "substring", QT_TRANSLATE_NOOP("gettextFromC", "with substring") },
+	{ FILTER_CONSTRAINT_EXACT, "exact", QT_TRANSLATE_NOOP("gettextFromC", "exactly") },
+};
+
+static struct range_mode_description {
+	filter_constraint_range_mode mode;
+	const char *s;
+	const char *translated;
+	const char *translated_date;
+} range_mode_descriptions[] = {
+	{ FILTER_CONSTRAINT_EQUAL, "equal", QT_TRANSLATE_NOOP("gettextFromC", "equal to"), QT_TRANSLATE_NOOP("gettextFromC", "at") },
+	{ FILTER_CONSTRAINT_LESS, "less", QT_TRANSLATE_NOOP("gettextFromC", "at most"), QT_TRANSLATE_NOOP("gettextFromC", "before") },
+	{ FILTER_CONSTRAINT_GREATER, "greater", QT_TRANSLATE_NOOP("gettextFromC", "at least"), QT_TRANSLATE_NOOP("gettextFromC", "after") },
+	{ FILTER_CONSTRAINT_RANGE, "range", QT_TRANSLATE_NOOP("gettextFromC", "in range"), QT_TRANSLATE_NOOP("gettextFromC", "in range") },
+};
+
+static const char *negate_description[2] {
+	QT_TRANSLATE_NOOP("gettextFromC", "is"),
+	QT_TRANSLATE_NOOP("gettextFromC", "is not"),
+};
+
+// std::size() is only available starting in C++17, so let's roll our own.
```

Comment:
I keep losing track... what C++ standard do we require these days? 11? 14?

Reply:
I think only C++11, at least we use only the C++11 version of the qOverload thing. I wouldn't rule out that we inadvertently introduced C++14 dependencies though.

That said, I would not be offended if you up the requirement to C++17 - it has some small but nifty syntactical sugar features, such as structured bindings.

Reply:
I don't think we have access to a C++17 compiler if we want to be useful AppImages
Similarly, we still support Ubuntu 16.04 and that might cause us some heartache with a C++17 compiler as well.

I believe we could introduce C++14 at this point (easiest way would be to add code that requires it and see what breaks) :-)

Reply:
> I believe we could introduce C++14 at this point (easiest way would be to add code that requires it and see what breaks) :-)

I went through the list of changes in C++14 and C++17 and frankly they are not really relevant to us. They concern mostly the standard-library and modern template programming, two things which we use only very lightly.

Reply:
pointless comment - turns out we no longer support Ubuntu 16.04 as that is based on Qt 5.5 and we have long left that behind. So C++14 might be an option